File: mvp/packages/consumer/src/state-mirror.ts
import type { SlopNode, PatchOp, SnapshotMessage, PatchMessage } from "@slop/types";

/**
 * Holds a local copy of a subscribed subtree and applies SLOP patches.
 * SLOP patches use node IDs in children segments, not array indices.
 */
export class StateMirror {
  private tree: SlopNode;
  private version: number;

  constructor(snapshot: SnapshotMessage) {
    this.tree = structuredClone(snapshot.tree);
    this.version = snapshot.version;
  }

  applyPatch(patch: PatchMessage): void {
    for (const op of patch.ops) {
      this.applyOp(op);
    }
    this.version = patch.version;
  }

  getTree(): SlopNode {
    return this.tree;
  }

  getVersion(): number {
    return this.version;
  }

  private applyOp(op: PatchOp): void {
    const segments = this.parsePath(op.path);
    if (segments.length === 0) return;

    switch (op.op) {
      case "add":
        this.applyAdd(segments, op.value);
        break;
      case "remove":
        this.applyRemove(segments);
        break;
      case "replace":
        this.applyReplace(segments, op.value);
        break;
    }
  }

  /**
   * Parse a SLOP patch path into segments.
   * "/children/todo-3/properties/done" → ["children", "todo-3", "properties", "done"]
   */
  private parsePath(path: string): string[] {
    return path.split("/").filter(Boolean);
  }

  /**
   * Navigate the tree to find the parent context and the last segment.
   * For children segments, we look up by node ID rather than array index.
   */
  private navigate(segments: string[]): { parent: any; key: string } | null {
    let current: any = this.tree;

    for (let i = 0; i < segments.length - 1; i++) {
      const seg = segments[i];
      if (seg === "children") {
        // Next segment is a child ID
        const childId = segments[i + 1];
        const child = (current.children as SlopNode[])?.find(
          (c) => c.id === childId
        );
        if (!child) return null;
        current = child;
        i++; // skip the ID segment
      } else if (seg === "properties" || seg === "meta" || seg === "affordances") {
        current = current[seg];
        if (current === undefined) return null;
      } else {
        current = current[seg];
        if (current === undefined) return null;
      }
    }

    return { parent: current, key: segments[segments.length - 1] };
  }

  private applyAdd(segments: string[], value: unknown): void {
    // Special case: adding a child node
    if (segments.length >= 2 && segments[segments.length - 2] === "children") {
      const parentSegments = segments.slice(0, -2);
      const parent = this.resolveNode(parentSegments);
      if (parent) {
        if (!parent.children) parent.children = [];
        parent.children.push(value as SlopNode);
      }
      return;
    }

    const target = this.navigate(segments);
    if (target) {
      target.parent[target.key] = value;
    }
  }

  private applyRemove(segments: string[]): void {
    // Special case: removing a child node
    if (segments.length >= 2 && segments[segments.length - 2] === "children") {
      const childId = segments[segments.length - 1];
      const parentSegments = segments.slice(0, -2);
      const parent = this.resolveNode(parentSegments);
      if (parent?.children) {
        parent.children = parent.children.filter((c: SlopNode) => c.id !== childId);
      }
      return;
    }

    const target = this.navigate(segments);
    if (target) {
      delete target.parent[target.key];
    }
  }

  private applyReplace(segments: string[], value: unknown): void {
    const target = this.navigate(segments);
    if (target) {
      target.parent[target.key] = value;
    }
  }

  /** Resolve a node given path segments (handling children/id pairs) */
  private resolveNode(segments: string[]): SlopNode | null {
    if (segments.length === 0) return this.tree;
    let current: SlopNode = this.tree;
    for (let i = 0; i < segments.length; i++) {
      const seg = segments[i];
      if (seg === "children") {
        const childId = segments[i + 1];
        const child = current.children?.find((c) => c.id === childId);
        if (!child) return null;
        current = child;
        i++;
      }
    }
    return current;
  }
}

File: packages/python/slop-ai/src/slop/state_mirror.py
"""StateMirror — maintains local tree state from snapshots and patches."""

from __future__ import annotations

import copy
from typing import Any

from .types import SlopNode, PatchOp


class StateMirror:
    """Mirrors a remote SLOP tree, applying snapshot and patch messages."""

    def __init__(self, snapshot: dict[str, Any]) -> None:
        self._tree = SlopNode.from_dict(copy.deepcopy(snapshot["tree"]))
        self._version: int = snapshot["version"]

    def apply_patch(self, patch: dict[str, Any]) -> None:
        """Apply a patch message (list of ops) to the local tree."""
        for op_data in patch["ops"]:
            op = PatchOp.from_dict(op_data) if isinstance(op_data, dict) else op_data
            self._apply_op(op)
        self._version = patch["version"]

    def get_tree(self) -> SlopNode:
        return self._tree

    def get_version(self) -> int:
        return self._version

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _apply_op(self, op: PatchOp) -> None:
        segments = [s for s in op.path.split("/") if s]
        if not segments:
            return
        if op.op == "add":
            self._apply_add(segments, op.value)
        elif op.op == "remove":
            self._apply_remove(segments)
        elif op.op == "replace":
            self._apply_replace(segments, op.value)

    def _navigate(self, segments: list[str]) -> tuple[Any, str] | None:
        """Walk *segments* down the tree, returning (parent, final_key).

        ``children`` segments resolve by child ID, not array index.
        ``meta`` navigates into a dataclass (attribute access).
        ``properties`` navigates into a dict (key access).
        """
        current: Any = self._tree
        i = 0
        while i < len(segments) - 1:
            seg = segments[i]
            if seg == "children":
                child_id = segments[i + 1]
                child = _find_child(current, child_id)
                if child is None:
                    return None
                current = child
                i += 2
            elif seg == "meta":
                current = getattr(current, "meta", None)
                if current is None:
                    return None
                i += 1
            elif seg == "properties":
                current = getattr(current, "properties", None) if hasattr(current, "properties") else None
                if current is None:
                    return None
                i += 1
            elif seg == "affordances":
                current = getattr(current, "affordances", None) if hasattr(current, "affordances") else None
                if current is None:
                    return None
                i += 1
            else:
                # generic attribute / key access
                if isinstance(current, dict):
                    current = current.get(seg)
                else:
                    current = getattr(current, seg, None)
                if current is None:
                    return None
                i += 1
        return (current, segments[-1])

    def _apply_add(self, segments: list[str], value: Any) -> None:
        # Adding a child node
        if len(segments) >= 2 and segments[-2] == "children":
            parent = self._resolve_node(segments[:-2])
            if parent is not None:
                if parent.children is None:
                    parent.children = []
                child = SlopNode.from_dict(value) if isinstance(value, dict) else value
                parent.children.append(child)
            return

        target = self._navigate(segments)
        if target is not None:
            parent, key = target
            if isinstance(parent, dict):
                parent[key] = value
            else:
                setattr(parent, key, value)

    def _apply_remove(self, segments: list[str]) -> None:
        # Removing a child node by ID
        if len(segments) >= 2 and segments[-2] == "children":
            child_id = segments[-1]
            parent = self._resolve_node(segments[:-2])
            if parent is not None and parent.children is not None:
                parent.children = [c for c in parent.children if c.id != child_id]
            return

        target = self._navigate(segments)
        if target is not None:
            parent, key = target
            if isinstance(parent, dict):
                parent.pop(key, None)
            else:
                setattr(parent, key, None)

    def _apply_replace(self, segments: list[str], value: Any) -> None:
        target = self._navigate(segments)
        if target is not None:
            parent, key = target
            if isinstance(parent, dict):
                parent[key] = value
            else:
                setattr(parent, key, value)

    def _resolve_node(self, segments: list[str]) -> SlopNode | None:
        """Walk segments to find a SlopNode (only ``children`` hops)."""
        if not segments:
            return self._tree
        current = self._tree
        i = 0
        while i < len(segments):
            if segments[i] == "children":
                if i + 1 >= len(segments):
                    return None
                child = _find_child(current, segments[i + 1])
                if child is None:
                    return None
                current = child
                i += 2
            else:
                i += 1
        return current


def _find_child(node: SlopNode, child_id: str) -> SlopNode | None:
    if node.children is None:
        return None
    for c in node.children:
        if c.id == child_id:
            return c
    return None
